vyper/signatures/function_signature.py

```diff
@@ -136,28 +139,41 @@ def from_definition(cls, code, sigs=None, custom_units=None, contract_def=False,
         method_id = fourbytes_to_int(sha3(bytes(sig, 'utf-8'))[:4])
         return cls(name, args, output_type, const, payable, private, sig, method_id, custom_units)
 
-    def _generate_output_abi(self):
+    def _generate_output_abi(self, custom_units_descriptions={}):
         t = self.output_type
-
         if not t:
             return []
         elif isinstance(t, TupleType):
-            res = [canonicalize_type(x) for x in t.members]
+            res = [(canonicalize_type(x), print_unit(unit_from_type(x), custom_units_descriptions)) for x in t.members]
         else:
-            res = [canonicalize_type(t)]
+            res = [(canonicalize_type(t), print_unit(unit_from_type(t), custom_units_descriptions))]
+
+        abi_outputs = [{"type": x, "name": "out", "unit": unit} for x, unit in res]
+
+        for abi_output in abi_outputs:
+            delete_unit_if_empty(abi_output)
 
-        return [{"type": x, "name": "out"} for x in res]
+        return abi_outputs
 
-    def to_abi_dict(self):
-        return {
+    def to_abi_dict(self, custom_units_descriptions={}):
```

Comment:
Please use custom_units_descriptions=None and then `if custom_units_descriptions is None ... ` instead. https://docs.quantifiedcode.com/python-anti-patterns/correctness/mutable_default_value_as_argument.html

Mutable default parameters can cause a lot of havoc ;)